File: npm-package/android/src/main/java/expo/modules/walletcore/WalletCoreModule.kt
package expo.modules.walletcore

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.walletcore.chains.EvmSigning
import expo.modules.walletcore.chains.SolanaSigning
import wallet.core.jni.AnyAddress
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.Mnemonic

// Chain-agnostic Expo Module bridging the kyehyukahn/wallet-core fork to
// React Native. Per-chain signing methods are appended in the same
// `definition()` body by calling static helpers in chains/<Chain>Signing.kt.

class WalletCoreModule : Module() {
    companion object {
        // wallet-core.aar bundles libTrustWalletCore.so for every ABI under
        // jni/<abi>/ — gradle's AGP repacks it into the consumer APK's
        // lib/<abi>/ — but the AAR ships zero classes that call
        // System.loadLibrary, so without this static init the very first
        // call into any wallet.core.jni.* class throws UnsatisfiedLinkError
        // (the JVM resolves Java_wallet_core_jni_<X>_native<...> against a
        // library that was never loaded). Expo Modules autolinking
        // instantiates WalletCoreModule during app startup, the JVM resolves
        // the class, this companion init fires, and the .so is available
        // for every subsequent JNI call.
        //
        // iOS does not need a symmetric step because Apple's static
        // linker resolves the xcframework symbols at link time.
        init {
            System.loadLibrary("TrustWalletCore")
        }
    }

    override fun definition() = ModuleDefinition {
        Name("WalletCoreModule")

        // ── chain-agnostic primitives ──

        Function("isValidMnemonic") { mnemonic: String ->
            Mnemonic.isValid(mnemonic)
        }

        // strength: 128 → 12 words, 256 → 24 words.
        AsyncFunction("createWallet") { strength: Int ->
            val wallet = HDWallet(strength, "")
            mapOf("mnemonic" to wallet.mnemonic())
        }

        // Derive an address for any registered CoinType + derivation path.
        AsyncFunction("deriveAddress") { mnemonic: String, coin: Int, path: String ->
            val coinType = CoinType.createFromValue(coin)
                ?: throw WalletCoreError.UnsupportedCoinType(coin)
            val wallet = HDWallet(mnemonic, "")
            val privateKey = wallet.getKey(coinType, path)
            // PrivateKey.getPublicKey(coinType) picks the right curve for the
            // coin internally. CoinTypeConfiguration does not expose
            // getPublicKeyType.
            val publicKey = privateKey.getPublicKey(coinType)
            AnyAddress(publicKey, coinType).description()
        }

        // ── EVM (chains/EvmSigning.kt) ──

        AsyncFunction("evmSignTransaction") { tx: Map<String, Any?>, mnemonic: String, path: String ->
            EvmSigning.signTransaction(tx, mnemonic, path)
        }

        // ── Solana (chains/SolanaSigning.kt) ──

        AsyncFunction("solanaSignTransfer") { tx: Map<String, Any?>, mnemonic: String, path: String ->
            SolanaSigning.signTransfer(tx, mnemonic, path)
        }
    }
}

sealed class WalletCoreError(message: String) : Exception(message) {
    class InvalidMnemonic : WalletCoreError("invalid BIP39 mnemonic")
    class UnsupportedCoinType(coin: Int) : WalletCoreError("unsupported CoinType id: $coin")
    class MissingField(field: String) : WalletCoreError("missing required field: $field")
    class InvalidField(field: String, reason: String) : WalletCoreError("invalid field $field: $reason")
    class ConflictingGasFields : WalletCoreError("conflicting gas fields: specify gasPrice OR max*PerGas, not both")
    class MissingGasFields : WalletCoreError("missing gas fields: provide either gasPrice or max*PerGas")
    class UnsupportedTransactionType(type: Int) : WalletCoreError("unsupported EVM transaction type: $type")
}

File: npm-package/ios/WalletCoreModule.swift
import ExpoModulesCore

// Chain-agnostic Expo Module bridging the kyehyukahn/wallet-core fork to
// React Native. Per-chain signing methods (evm*, solana*) are appended in
// the same `definition()` body — Expo's ModuleDefinition DSL must run in
// one place, so each chain provides static helpers in chains/<Chain>Signing.swift
// that this file calls from inside the DSL block.

public class WalletCoreModule: Module {
  public func definition() -> ModuleDefinition {
    Name("WalletCoreModule")

    // ── chain-agnostic primitives ──

    // Mnemonic validation is pure (no derivation), so it's a synchronous
    // Function rather than AsyncFunction.
    Function("isValidMnemonic") { (mnemonic: String) -> Bool in
      Mnemonic.isValid(mnemonic: mnemonic)
    }

    // strength: 128 → 12 words, 256 → 24 words. wallet-core enforces.
    AsyncFunction("createWallet") { (strength: Int) -> [String: String] in
      guard let wallet = HDWallet(strength: Int32(strength), passphrase: "") else {
        throw WalletCoreError.hdWalletCreationFailed
      }
      return ["mnemonic": wallet.mnemonic]
    }

    // Derive an address for any registered coin type + derivation path.
    // Per-chain subpaths (evm, solana) call this with their CoinType numeric.
    AsyncFunction("deriveAddress") { (mnemonic: String, coin: Int, path: String) -> String in
      guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
        throw WalletCoreError.invalidMnemonic
      }
      guard let coinType = CoinType(rawValue: UInt32(coin)) else {
        throw WalletCoreError.unsupportedCoinType(coin)
      }
      let privateKey = wallet.getKey(coin: coinType, derivationPath: path)
      let pubKeyType = CoinTypeConfiguration.getPublicKeyType(coin: coinType)
      let publicKey = privateKey.getPublicKeyByType(pubkeyType: pubKeyType)
      return AnyAddress(publicKey: publicKey, coin: coinType).description
    }

    // ── EVM (chains/EvmSigning.swift) ──

    AsyncFunction("evmSignTransaction") { (tx: [String: Any], mnemonic: String, path: String) -> String in
      try EvmSigning.signTransaction(tx: tx, mnemonic: mnemonic, derivationPath: path)
    }

    // ── per-chain signing methods are added by chains/<Chain>Signing.swift ──
    // commit 7: Solana (solanaSignTransfer)
  }
}

enum WalletCoreError: Error, LocalizedError {
  case hdWalletCreationFailed
  case invalidMnemonic
  case unsupportedCoinType(Int)
  case missingField(String)
  case invalidField(String, String)
  case conflictingGasFields
  case missingGasFields
  case unsupportedTransactionType(Int)

  var errorDescription: String? {
    switch self {
    case .hdWalletCreationFailed:
      return "HDWallet creation failed (invalid strength?)"
    case .invalidMnemonic:
      return "invalid BIP39 mnemonic"
    case .unsupportedCoinType(let n):
      return "unsupported CoinType id: \(n)"
    case .missingField(let f):
      return "missing required field: \(f)"
    case .invalidField(let f, let reason):
      return "invalid field \(f): \(reason)"
    case .conflictingGasFields:
      return "conflicting gas fields: specify gasPrice OR max*PerGas, not both"
    case .missingGasFields:
      return "missing gas fields: provide either gasPrice or max*PerGas"
    case .unsupportedTransactionType(let t):
      return "unsupported EVM transaction type: \(t)"
    }
  }
}
